spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/IcebergSource.java

```diff
@@ -115,10 +133,28 @@ public Table getTable(StructType schema, Transform[] partitioning, Map<String, S
     if (catalogAndIdentifier.catalog().name().equals("spark_catalog") &&
         !(catalogAndIdentifier.catalog() instanceof SparkSessionCatalog)) {
       // catalog is a session catalog but does not support Iceberg. Use Iceberg instead.
+      Identifier ident = catalogAndIdentifier.identifier();
       return new Spark3Util.CatalogAndIdentifier(catalogManager.catalog(DEFAULT_CATALOG_NAME),
-          catalogAndIdentifier.identifier());
-    } else {
+          newIdentifier(ident, selector));
+    } else if (snapshotId == null && asOfTimestamp == null) {
       return catalogAndIdentifier;
+    } else {
+      CatalogPlugin catalog = catalogAndIdentifier.catalog();
+      Identifier ident = catalogAndIdentifier.identifier();
+      return new Spark3Util.CatalogAndIdentifier(catalog,
+          newIdentifier(ident, selector));
+    }
+  }
+
+  private Identifier newIdentifier(Identifier ident, String newName) {
```

Comment:
This isn't very descriptive. I would rather call it `identifierWithSelector` and rename `newName` to `selector`.

Reply:
Renamed.